gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+        return command.getPermissionChecks().keySet().stream()
+                .collect(Collectors.toMap(
+                        check -> check,
+                        check -> false,
+                        (oldValue, newValue) -> oldValue,
+                        LinkedHashMap::new));
+    }
+
+    /**
+     * Groups permission checks by entity ID and policy resource.
+     *
+     * @param command the {@link CheckPermissions} command.
+     * @return a map grouped by entity ID, containing permission check wrappers.
+     */
+    private Map<String, Map<String, PermissionCheckWrapper>> groupByEntityAndPolicyResource(
+            final CheckPermissions command) {
+        return command.getPermissionChecks().entrySet().stream()
+                .collect(Collectors.groupingBy(
+                        entry -> entry.getValue().getEntityId(),
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                entry -> new PermissionCheckWrapper(entry.getValue())
+                        )
+                ));
+    }
+
+    /**
+     * Retrieves or sets the {@link PolicyId} for a given entity.
+     * <p>
+     * If the entity is a policy resource, it directly sets the {@link PolicyId} for the corresponding wrappers.
+     * If the entity is not a policy resource, it retrieves the {@link PolicyId} from the entity using the
+     *
+     * @param entityId the ID of the entity for which the {@link PolicyId} is being retrieved or set.
+     * @param wrappers the wrappers containing the permission checks for the entity.
+     * @param command the {@link CheckPermissions} command containing the headers for the request.
+     * @return a {@link CompletableFuture} that completes when the {@link PolicyId} is set for all wrappers.
+     */
+    private CompletableFuture<Void> retrieveOrSetPolicyId(final String entityId,
+            final Map<String, PermissionCheckWrapper> wrappers,
+            CheckPermissions command) {
+        if (wrappers.values().stream().findFirst().get().getPermissionCheck().isPolicyResource()) {
+            wrappers.values().forEach(wrapper -> wrapper.setPolicyId(PolicyId.of(entityId)));
+            return CompletableFuture.completedFuture(null);
+        } else {
+            return retrievePolicyIdForEntity(entityId, false, command.getDittoHeaders())
+                    .thenAccept(policyId -> wrappers.values().forEach(wrapper -> wrapper.setPolicyId(policyId)))
+                    .toCompletableFuture();
+        }
+    }
+
+    /**
+     * Handles the completion of policy retrieval by checking the permissions for each policy.
+     *
+     * @param command the {@link CheckPermissions} command.
+     * @param permissionResults the map to store the permission check results.
+     * @param groupedByEntityId the permission checks grouped by entity ID.
+     */
+    private void handlePolicyRetrievalCompletion(final CheckPermissions command,
```

Comment:
This method should return a `CompletionStage<Void>`, which the caller would invoke with `thenCompose` - in order to "wait" for its completion before e.g. completing itself.